Fixing error and added groupFC ordering on plotAbundanceTest

halaszlaszlo opened the pull request:
Fixed error caused by large numbers in pseudoCount function. Added ordering for plotAbundanceTest by groupFC.

halaszlaszlo committed:
Update lotOfCells.R

Fix: Handle integer overflow in pseudoCount function

halaszlaszlo committed:
Update plotAbundanceTest.R

Added groupFC ordering
Removed extra labels from the plot

halaszlaszlo committed:
Update plotAbundanceTest.R

Fixed ordering by groupFC

OscarGVelasco commented:
Hi @halaszlaszlo , thanks a lot for contributing!
I believe the ordered values on the abundance plot are much nicer and easier to interpret.

Just out of curiosity, what was the error caused by large numbers in pseudoCount function? I have made a small test with large numbers and the results seem consistent between "counts*counts" and "counts^2". As I use this function for other projects, I would like to know more.

Regarding the aesthetics of the plot, I would like to keep the theme_minimal() only (not adding a ggplot2::theme_bw()) for clearness and consistency with other plots. Also, please keep the title and axis labels, as they are informative for the user (the labels indicate the variables used by the user). In general, all these modifications can be done by the user by just modifying the returned ggplot object. Also, in future occasions, please state all these aesthetic changes in the pull request, as they modify substantially the output.

Finally, I have check on a fresh R session and encountered this error (I believe tibble needs to be added as a dependency):

```
Only 2 groups detected.
Computing Fold Change proportion over covariables for groups: tLung vs nLung
Additional sub-level for testing: Patient id
- Starting montecarlo simulation of fold changes
Error in h(simpleError(msg, call)) : 
  error in evaluating the argument 'object' in selecting a method for function 'show': could not find function "rownames_to_column"

```

Again, thanks a lot for contributing.

halaszlaszlo committed:
Update plotAbundanceTest.R

Kept groupFC ordering, removed tibble dependency and set plot aesthetics to original.

halaszlaszlo commented:
Thanks for the great package! @OscarGVelasco 

I changed back the aesthetics to original; removed tibble dependency and kept the groupFC ordering for _plotAbundanceTest_.

The error was indeed weird. I had 88 classLabels (>1M) and only one label (n=176215 [condA = 127937; condB = 48278]) caused the error and produced NAs ("_In counts * counts : NAs produced by integer overflow_").

Changing: "counts*counts" to "counts^2" solved it and no errors. Not sure what is the exact reason for error. 

```
lotOfCells(scObject = metadata_df,
                   main_variable = "condition",
                   subtype_variable = "granular_annotation",
                   sample_id = "sample_id",
                   permutations = 10000,
                   labelOrder = c("condA", "condB"),
                   parallel = TRUE) 
```

```
R version 4.3.2 (2023-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Sonoma 14.5

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib 
LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

time zone: America/New_York
tzcode source: internal

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] LotOfCells_0.2.0 lubridate_1.9.3  forcats_1.0.0    stringr_1.5.1    dplyr_1.1.4      purrr_1.0.2      readr_2.1.5      tidyr_1.3.1     
 [9] tibble_3.2.1     ggplot2_3.5.1    tidyverse_2.0.0 

loaded via a namespace (and not attached):
 [1] SummarizedExperiment_1.32.0 gtable_0.3.5                rstatix_0.7.2               Biobase_2.62.0             
 [5] lattice_0.22-6              tzdb_0.4.0                  vctrs_0.6.5                 tools_4.3.2                
 [9] bitops_1.0-7                generics_0.1.3              stats4_4.3.2                parallel_4.3.2             
[13] fansi_1.0.6                 pkgconfig_2.0.3             Matrix_1.6-5                S4Vectors_0.40.2           
[17] lifecycle_1.0.4             GenomeInfoDbData_1.2.11     compiler_4.3.2              munsell_0.5.1              
[21] codetools_0.2-20            carData_3.0-5               GenomeInfoDb_1.38.8         RCurl_1.98-1.14            
[25] car_3.1-2                   pillar_1.9.0                ggpubr_0.6.0                crayon_1.5.3               
[29] BiocParallel_1.36.0         SingleCellExperiment_1.24.0 DelayedArray_0.28.0         abind_1.4-5                
[33] tidyselect_1.2.1            stringi_1.8.4               reshape2_1.4.4              grid_4.3.2                 
[37] colorspace_2.1-0            cli_3.6.3                   SparseArray_1.2.4           magrittr_2.0.3             
[41] S4Arrays_1.2.1              utf8_1.2.4                  broom_1.0.6                 withr_3.0.0                
[45] backports_1.5.0             scales_1.3.0                timechange_0.3.0            XVector_0.42.0             
[49] matrixStats_1.3.0           gridExtra_2.3               ggsignif_0.6.4              hms_1.1.3                  
[53] GenomicRanges_1.54.1        IRanges_2.36.0              rlang_1.1.4                 Rcpp_1.0.12                
[57] glue_1.7.0                  BiocGenerics_0.48.1         rstudioapi_0.16.0           R6_2.5.1                   
[61] plyr_1.8.9                  MatrixGenerics_1.14.0       zlibbioc_1.48.2    
```

OscarGVelasco commented:
Thanks a lot for this @halaszlaszlo, I was able to reproduce the error. It turns out that table() return integers, and then integer*integer can incur on _integer overflow_, and apparently integer^2 is cast as numeric before squaring the value.
Again, thanks for your contribution!
